nncf/common/pruning/utils.py

```diff
@@ -636,4 +645,44 @@ def identity_mask_propagation(node: NNCFNode, graph: NNCFGraph) -> None:
         # In case for disconnected NNCFGraph
         input_masks = [None]
     assert len(input_masks) == 1
-    node.data['output_mask'] = input_masks[0]
+    input_mask = input_masks[0]
+
+    if isinstance(input_mask, list):
+        input_mask = find_input_mask_for_node(input_mask, node)
+    node.data['output_mask'] = input_mask
+
+
+def match_multiple_output_masks(output_masks: List[SymbolicMask], output_edges: List[NNCFGraphEdge],
+                                chunk_axis: int) -> dict:
+    """
+    Match multiple input mask to each next nodes.
+
+    :param output_masks: Given output masks.
+    :return: Matched output mask for each next node.
+    """
+    output_shapes = [edge.tensor_shape[chunk_axis] for edge in output_edges]
+    next_nodes = [edge.to_node for edge in output_edges]
+    result_masks = {node.node_name: None for node in next_nodes}
+
+    if len(set(output_shapes)) == 1:
+        for i, split_mask in enumerate(output_masks):
+            result_masks[next_nodes[i].node_name] = split_mask
+    else:
+        for split_mask in output_masks:
+            idx = output_shapes.index(split_mask.shape[0])
+            result_masks[next_nodes[idx].node_name] = split_mask
+            # update already matched
+            output_shapes[idx]=None
+
+    return list(result_masks.items())
```

Comment:
I don't feel comfortable to match output masks to input nodes by shape, it could lead to mistakes in case shapes are equal and masks values are different on each branch. Can we keep index of input node for masks instead of mapping?

Reply:
The mapping process is absolutely necessary. After spliting the mask, we can not miss the process of finding the right mask for which node. It doesn't seem to matter which node the mask we found fits, whether it is index or node name. When we look at the shape of the mask, we don't look at the entire shape, we only look at the axis that we split.